File: leetcode/golang/54_spiral_matrix.go
func spiralOrder(matrix [][]int) []int {
    if len(matrix) == 0 || len(matrix[0]) == 0 {
        return []int{}
    }
    l, r, t, b := 0, len(matrix[0]) - 1, 0, len(matrix) - 1
    res := make([]int, 0)
    for {
        for i := l; i <= r; i++ {
            res = append(res, matrix[t][i])
        }
        t++
        if t > b {
            break
        }
        for i := t; i <= b; i++ {
            res = append(res, matrix[i][r])
        }
        r--
        if r < l {
            break
        }
        for i := r; i >= l; i-- {
            res = append(res, matrix[b][i])
        }
        b--
        if b < t {
            break
        }
        for i := b; i >= t; i-- {
            res = append(res, matrix[i][l])
        }
        l++
        if l > r {
            break
        }
    }
    return res
}

File: leetcode/c++/54_spiral_matrix.cpp
// https://leetcode.cn/problems/spiral-matrix/description

#include  <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if (matrix.empty()) return {};
        int l = 0, r = matrix[0].size() - 1, t = 0, b = matrix.size() - 1;
        vector<int> res;
        while (true) {
            for (int i = l; i <= r; i++) res.push_back(matrix[t][i]);
            if (++t > b) break;
            for (int i = t; i <= b; i++) res.push_back(matrix[i][r]);
            if (--r < l) break;
            for (int i = r; i >= l; i--) res.push_back(matrix[b][i]);
            if (--b < t) break;
            for (int i = b; i >= t; i--) res.push_back(matrix[i][l]);
            if (++l > r) break;
        }
        return res;
    }
};

// 时间复杂度为O(m*n)，空间复杂度为O(1)